chips/rp2040/src/usb.rs

```diff
@@ -0,0 +1,2532 @@
+                        self.handle_epdata_in(15);
+                    }
+                    self.registers
+                        .buff_status
+                        .modify(BUFF_STATUS::EP15_IN::CLEAR);
+                }
+                31 => {
+                    if self.registers.buff_status.is_set(BUFF_STATUS::EP15_OUT) {
+                        self.handle_epdata_out(15);
+                    }
+                    self.registers
+                        .buff_status
+                        .modify(BUFF_STATUS::EP15_OUT::CLEAR);
+                }
+                _ => {}
+            }
+        }
+        self.registers.buff_status.set(0);
+    }
+
+    fn process_requests(&self) {
+        for (endpoint, desc) in self.descriptors.iter().enumerate() {
+            if desc.request_transmit_in.take() {
+                if endpoint == 0 {
+                    self.transmit_in_ep0();
+                } else {
+                    self.transmit_in(endpoint);
+                }
+            }
+            if desc.request_transmit_out.take() {
+                if endpoint == 0 {
+                    self.transmit_out_ep0();
+                } else {
+                    self.transmit_out(endpoint);
+                }
+            }
+        }
+    }
+
+    fn handle_epdata_out(&self, ep: usize) {
+        let (transfer_type, in_state, out_state) = self.descriptors[ep].state.get().bulk_state();
+        assert!(out_state.is_some());
+
+        // We need to read the size at this point in the process. At this point
+        // the USB hardware has received the data, but we need to
+        // copy the data to memory. Later on the EPOUT.SIZE register can
+        // be overwritten, particularly if the host is sending OUT
+        // transactions quickly.
+        let ep_size = self.dpsram.ep_buf_ctrl[ep]
+            .ep_out_buf_ctrl
+            .read(EP_BUFFER_CONTROL::TRANSFER_LENGTH0);
+
+        match out_state.unwrap() {
+            BulkOutState::Init => {
+                // The endpoint is ready to receive data. Request a transmit_out.
+                self.descriptors[ep].request_transmit_out.set(true);
+            }
+            BulkOutState::OutDelay => {
+                // The endpoint will be resumed later by the client application with transmit_out().
+            }
+            BulkOutState::OutData { size: _ } => {
+                self.descriptors[ep].request_transmit_out.set(true);
+            }
+        }
+        // Indicate that the endpoint now has data available.
+        self.descriptors[ep].state.set(EndpointState::Bulk(
+            transfer_type,
+            in_state,
+            Some(BulkOutState::OutData { size: ep_size }),
+        ));
+    }
+
+    fn handle_epdata_in(&self, endpoint: usize) {
+        let (transfer_type, in_state, out_state) =
+            self.descriptors[endpoint].state.get().bulk_state();
+        assert!(in_state.is_some());
```

Comment:
same here

Reply:
We just followed the nrf implementation. I think this modification will require more work and should be in an improvement PR.